### DISCUSSION TITLE
Why do plugins need to get physically drawn on the server computer.?

### DISCUSSION THREAD

**Discussion by johne53** [apohl79/audiogridder#447] [Ideas]
I've been trying Audiogridder for a few days and I'm hugely impressed - but I only realised today that each time I open a plugin on my client machine, it also gets drawn on the server machine (i.e. when using Windows...)

Even worse... if someone needs to do something on the server machine, they can't minimize the plugin or move it "out of the way" without it causing problems on the client machine. And (presumably) it won't be possible to use a large number of plugins if the server machine only has a small desktop resolution..?

I just wondered if there are any plans to overcome those limitations in the future? e.g. could the plugin get "drawn" onto a server window that's minimized?

**Comment by apohl79** [comment]
The remote UI rendering is based on screen capturing. That means, minimized windows can't be captured. Also one server instance will show only a single plugin UI at a time.

**Reply by johne53** [reply]
Wow, you're right - I hadn't noticed that, so far. Basically, if 2 plugins have got something that moves (let's say, meters) If I select one plugin, the other one's display immediately freezes. That's quite a big drawback... :-(

**Comment by bsjdsp** [comment]
Any plans for a better solution or is it best to run multiple servers to open multiple windows at the same time

**Reply by johne53** [reply]
FWIW a user on the Mixbus forum has been experimenting with this and it does work - except it's all a bit kludgy if the servers are sharing the same machine. I'm not very familiar with Mac but on Windows, the solution will probably involve drawing to Device Contexts, rather than relying on physical screen captures etc.

Note to the developers... please give this some serious thought. Audiogridder is a fantastic idea - but this drawing limitation is a massive let-down at the moment :-(

**Reply by Clintcinnati** [reply]
That Mixbus user is me!
I exploit the fact that the serving machine has the windows of the plugins open, and don’t use screen capture at all, so more the bandwidth between the server and the served rather than the GUI.
Part of my problem (and anyone who has tried Mixbus has likely noticed) is screen space is dominated by the Mixbus Mixer when in mix mode, so having the plugins on a separate screen is almost essential.
The thing to overcome, is on the server computer, one must spawn multiple instances of the server so there can be more than one gui open and active.  So say you want to run 8 plugins from the server to the daw, and you’d like to be able to see all those 8 gui open at once and responsive to mouse commands during the session.
As I said, that means there needs to be 8 servers serving, because if you just use one server, then only one GUI at a time on the server machine will be visible.

I was able to get multiple servers going before by simply making 8 copies of the server app and opening all 8 of them, but later versions of AGridder would not work that way because they all try to use the same port.  Or so my limited understanding allowed.

By conversing with this forum, I was told that it will work to have multiple server instances, so long as the servers are given the appropriate argument via the command line.

So I wound up writing a script in bash that does just that, and made it into an executable by googling how to do that on a mac.

When I show you my bash command, I warn you that I know there has got to be a much more elegant way of writing what I did, but it does work every time.

Here is my script, which could be way more elegant if I knew what I was doing-

clint@Lefty ~ % cat gridder
!#/bin/bash
open -n /Applications/AudioGridderServer.app --args -id 1;open -n /Applications/AudioGridderServer.app --args -id 2;open -n /Applications/AudioGridderServer.app --args -id 3;open -n /Applications/AudioGridderServer.app --args -id 4;open -n /Applications/AudioGridderServer.app --args -id 5;open -n /Applications/AudioGridderServer.app --args -id 6;open -n /Applications/AudioGridderServer.app --args -id 7;open -n /Applications/AudioGridderServer.app --args -id 8
clint@Lefty ~ % 

Naturally, your directory location is not going to be the same as mine nor is your computer name etc.  And I just know that instead of me repeating the same line 8 times, I could have used a way to increment, but in my case it was easier to paste the initial command 8 times and edit each command to increment.

I used some mac stuff to make that command into an app, so when I boot up to start a session, I fire up the server computer first, then click the app that really just opens the terminal and runs that above pasted command.  
Eight audiogridder icons appear in the menu bar of my server mac, and before I turn on the daw computer, I open the server settings window on each of the eight and change if needed the Server ID number to match the Server ID (commandline overide) server ID number.
Also I make sure for my purposes to select Disabled in Screen Capturing Mode.  I do this for each of the eight servers, the  push the save button at the bottom of each.

Be sure that you don’t have wifi turned on on the server, you need ethernet or better wired connection to the daw computer.
If wifi is on, in the case of my macs, the daw will try to use wifi to connect to the server and it’s a freakin mess.
Wifi being active on the daw computer is fine tho.

After firing up
the daw computer, and adding audiogridder plugins into the session, you’ll hopefully see when you get to the list of servers to connect to, a list of eight servers with the name of your server computer preceding each.  
 I know it seems like a chore to set up, and if I knew more about what I was doing writing that command, I doubt I’d have to do the steps involving opening the server settings windows.

Now with no screen capture going on, that means I use two mice, one for the daw and one for
the server.  

The performance gains are enormous!  My imacs are late 2012 models.  If I have several plugins open on the daw mac, along with the scrolling editor and 60 channels of the mixer open, the response to changing things on the screen is excruciatingly slow, like half second or more.  Mixbus as a daw is wonderful but with all
those channels open, each is a burden with the included, always on, compressors and eq in each channel, along with meters etc.

Taking the graphical load off of the DAW mac for the GUI of all the compressors and plugins that have FFT displays etc, it’s like going from a propellor plane to a jet.

I do hope that this mode of non-screen capture, multi server from the single connected server computer will become a recognized official
mode of operation or at least an option, I’ve tried the screen capture way and it just clutters my daw screen and slows things down.
Having eight compressors open and visible on a separate screen with the meters actually responding and controls
working, it’s just amazing.

**Reply by johne53** [reply]
Interesting stuff!! And in fact it prompted me to check a couple of things here... all my computers currently connect to my router via wired Ethernet and all of them show a transmission speed of 100Mb/s - but my newest machine offers 802.11ac wifi. After a quick check online, it seems that's more than a dozen times faster - offering a whopping 1300Mb/s!! So in theory, things would speed up massively if I just unplugged the Ethernet and switched to using wifi.

But of course, that all depends on the router's wifi adapter and I'm not sure how to check it. It's the standard router provided by my ISP (TalkTalk) and knowing my luck, it'll probably be a slow one...

[Edit...] Darn it!!  I just checked on the back of the router and it's only 11Mb/s :-(

**Comment by johne53** [comment]
Also... does this limitation only affect the Windows server or is MacOS also affected?

IIUC the Audiogridder server isn't available yet for Linux - but there's something in the back of my mind that Linux can support multiple simultaneous desktops (i.e. a user can switch between different ones without needing to reboot). Maybe that could offer a workaround for this problem in the future?

Also... Windows offers something called **Device Contexts** which can be used to implement a drawing surface - but without needing to physically draw anything onto the screen. Maybe that could also offer a potential solution at some point? I'm sure it's possible to retrieve the contents of a Device Context without any need for a physical screen capture.

**Comment by johne53** [comment]
Even if my suggestions aren't the best way forward... do the developers foresee a time when this restriction will get removed eventually? (i.e. a time when multiple plugins would be able to update their GUI's simultaneously ?)

**Reply by apohl79** [reply]
Hey, I'm thinking about this for a while already. To me the best approach would be to use some kind of a virtual desktop approach, that works for macOS and Windows similarly. So in my mind AG should create a virtual desktop that is not attached to a screen and not visible to a user, but it has to allow to be captured via gdi on Windows and avfoundation on Mac. Tbh, I don't know if this is possible. Research is required.

I'm not sure the approach you are proposing would work, as plugins can create windows, render tool tips, popup menus, can use opengl etc. I think the only way to transfer the full UI consistently is to capture the screen.

I will create a feature request ticket. But i can't provide any ETA.

**Reply by johne53** [reply]
That'd be fantastic.... Audiogridder is a great idea but this GUI limitation does kinda let it down - in fact to be honest, it gives it an almost amateurish feel :-(

And I'm not saying that in any nasty way - it's just intended as an observation.

I'm a programmer myself (tho' mostly involved with debugging these days) so I do appreciate all the hard work that's gone into Audiogridder !!

**Reply by apohl79** [reply]
Tbh - I never felt like there needs to keep more than one plugin UI in sync and it's nothing someone ever asked before. AG has to make compromises to work around plugin frameworks that are not designed for its use case, and the UI is clearly a best effort thing. The main purpose of this tool - for me - was to offload the DSP processing part...

Now as you are a programmer, please feel welcome to contribute to the project! Any help on the coding side is very much appreciated! ;-)

**Reply by johne53** [reply]
I don't have much experience with graphics programming (it's mostly file conversions and debugging...)

I guess the UI stuff is something that's becoming ever more common. Many VST's have some kind of animated display these days - even if it's just a couple of audio meters.

**Comment by CaBuffalo** [comment]
Johne53, your router must be a 10/100f instead of 10/100/1000.

I highly recommend you to use LAN instead of Wifi.

For internet access, your wifi speed is pretty good. Using AG doesn't
require high speed internet speed. Lan speed is the key.

On Mon, Apr 5, 2021 at 4:39 AM johne53 ***@***.***> wrote:

> Interesting stuff!! And in fact it prompted me to check a couple of things
> here... all my computers currently connect to my router via wired Ethernet
> and all of them show a transmission speed of 100Mb/s - but my newest
> machine offers 802.11ac wifi. After a quick check online, it seems that's
> more than a dozen times faster - offering a whopping 1300Mb/s!! So in
> theory, things would speed up massively if I just unplugged the Ethernet
> and switched to using wifi.
>
> But of course, that all depends on the router's wifi adapter and I'm not
> sure how to check it. It's the standard router provided by my ISP
> (TalkTalk) and knowing my luck, it'll probably be a slow one... :-(
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apohl79/audiogridder/discussions/447#discussioncomment-570088>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARMAR6H7JXCX5F73KPEX2ITTHGOODANCNFSM4ZU7VF3A>
> .
>


-- 
Kevin Ma
Supported by Gmail.

**Comment by johne53** [comment]
Many thanks CaBuffalo - BTW, as a follow-up to **apohl79** I tried a few internet searches and it seems there are various packages available for providing Windows virtual desktops but AFAICT they all seem to be cloud based solutions. I can't find out if Windows offers virtual desktops internally. Presumably an internal solution is what you'd need for Audiogridder?

**Comment by johne53** [comment]
Hi **apohl79** - I don't want to sound discouraging but over the past few days I've done some reading about Virtual Desktops (in Windows). I was hoping to find some simple way to implement them but everything I've read makes it sound fearsomely complicated :-(

AFAICT Windows itself doesn't support running multiple desktops simultaneously and IIUC this means that to achieve virtual desktops, each desktop needs to run a complete underlying copy of Windows!! I'm guessing that's why they tend to get implemented as cloud-based solutions...

Fixing this in Audiogridder (i.e. the problem with multiple plugin GUI's) would be such a massive improvement so like I said, I don't want to discourage it.  But if you can get your head around Device Contexts, I've a feeling they'll offer a much quicker and simpler solution. A Device Context can implement a window (i.e. a drawing surface) but without it needing to be physically visible on screen. They've been around for a long time so it should also be easier to find support and examples etc.

in a way, I hope I'm totally wrong about Windows Virtual Desktops - especially if they're easy to implement in MacOS -but I just wanted to offer a heads-up...

**Reply by johne53** [reply]
Woohoo!! It looks like I might be wrong about this! Apparently, Windows does offer virtual desktops as part of something called Window Stations - which I must admit, I'd never heard of:- [https://docs.microsoft.com/en-us/windows/win32/winstation/window-stations](url)

In fact there's even a 'C' function called **CreateDesktop** which can create desktops programmatically (presumably multiple ones) and then associate them with the current Window Station:- [https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-createdesktopa](url)

[Edit...] BTW those links don't seem to work for me here (Microsoft Edge browser) but they do seem to work if I copy & paste them to a new browser page.

**Reply by apohl79** [reply]
I created ticket https://github.com/apohl79/audiogridder/issues/476 and will have a look once I get some capacity for this. Thx for sharing.

**Comment by johne53** [comment]
I guess there's not much happening with this - but...

> johne53 wrote:-
> IIUC the Audiogridder server isn't available yet for Linux - but there's something in the back of my mind that Linux can support multiple simultaneous desktops

I've heard rumours that the new Windows 11 can also support multiple simultaneous desktops - so does that offer a potential solution? In theory each Audiogridder plugin could occupy its own hidden desktop without hindering the server's main desktop.

**Reply by johne53** [reply]
I just wondered if there's been any progress with this?

**Comment by CaBuffalo** [comment]
AG needs to capture that area on server computer. So, it will be displayed.

On Sat, Oct 23, 2021 at 4:46 AM johne53 ***@***.***> wrote:

> I just wondered if there's been any progress with this?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/apohl79/audiogridder/discussions/447#discussioncomment-1524775>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARMAR6FJL26COEMLZ5P7V73UIKN75ANCNFSM4ZU7VF3A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Kevin Ma
Supported by Gmail.

**Comment by johne53** [comment]
What I'd heard was that with the introduction of Windows 11, Windows and Mac can now both support multiple simultaneous desktops. So does that offer a solution for the problem of AudioGridder only being able to display 1 x plugin in real time? i.e. could AudioGridder's server be adapted to make use of these multiple desktop features?

The fact that AudioGridder can only update 1 x plugin's display in real time is a massive limitation :-(

**Reply by apohl79** [reply]
I don't know. I have not looked at this and don't know how these OS API's work and if new desktops can be created/removed by an application as needed. But in general to make AG support capturing multiple windows at the same time requires some window management functionality to manage the available screen space, as windows should not overlap etc.

**Reply by johne53** [reply]
> 
> The fact that AudioGridder can only update 1 x plugin's display in real time is a massive limitation :-(
> 

I was just wondering if that limitation ever got fixed..?
